product/util.py

```diff
@@ -171,164 +211,197 @@ def diff_products(left_product, right_product, ignore_component_names=()):
     )
 
 
-@dataclasses.dataclass
-class ComponentDiff:
-    crefs_only_left: set = dataclasses.field(default_factory=set)
-    crefs_only_right: set = dataclasses.field(default_factory=set)
-    crefpairs_version_changed: set = dataclasses.field(default_factory=set)
-    names_only_left: set = dataclasses.field(default_factory=set)
-    names_only_right: set = dataclasses.field(default_factory=set)
-    names_version_changed: set = dataclasses.field(default_factory=set)
-
+def diff_components(
+    left_components: typing.Tuple[cm.Component],
+    right_components: typing.Tuple[cm.Component],
+    ignore_component_names=(),
+) -> ComponentDiff:
+    left_identities = {
+        c.identity() for c in left_components if c.name not in ignore_component_names
+    }
+    right_identities = {
+        c.identity() for c in right_components if c.name not in ignore_component_names
+    }
 
-def diff_components(left_components, right_components, ignore_component_names=()) -> ComponentDiff:
-    left_components = set(left_components)
-    right_components = set(right_components)
+    left_only_identities = left_identities - right_identities
+    right_only_identities = right_identities - left_identities
 
-    if left_components == right_components:
+    if left_only_identities == right_only_identities:
         return None # no diff
 
-    components_only_left = left_components - right_components
-    components_only_right = right_components - left_components
+    left_components = tuple((
+        c for c in left_components if c.identity() in left_only_identities
+    ))
+    right_components = tuple((
+        c for c in right_components if c.identity() in right_only_identities
+    ))
 
-    def find_changed_component(changed_component, components):
+    def find_changed_component(
+        changed_component: cm.Component,
+        components: typing.List[cm.Component],
+    ):
         for c in components:
-            if c.name() == changed_component.name():
+            if c.name == changed_component.name:
                 return (changed_component, c)
         return (changed_component, None) # no pair component found
 
-    components_with_changed_versions = FluentIterable(items=components_only_left) \
+    components_with_changed_versions = FluentIterable(items=left_components) \
```

Comment:
nit: maybe we should replace this `FluentIterable` (which I stole from guava^^) with a plain list-comprehension, wdyt?